src/devices/src/virtio/balloon/device.rs

```diff
@@ -0,0 +1,920 @@
+            error!("Failed to signal used queue: {:?}", e);
+            DeviceError::FailedSignalingUsedQueue(e)
+        })?;
+        Ok(())
+    }
+
+    pub fn update_num_pages(&mut self, num_pages: u32) {
+        self.config_space.num_pages = num_pages;
+    }
+
+    pub fn num_pages(&self) -> u32 {
+        self.config_space.num_pages
+    }
+
+    pub fn latest_stats(&self) -> &BalloonStats {
+        &self.latest_stats
+    }
+
+    pub(crate) fn stats_enabled(&self) -> bool {
+        self.stats_polling_interval_s > 0
+    }
+}
+
+impl VirtioDevice for Balloon {
+    fn device_type(&self) -> u32 {
+        TYPE_BALLOON
+    }
+
+    fn queues(&self) -> &[Queue] {
+        &self.queues
+    }
+
+    fn queues_mut(&mut self) -> &mut [Queue] {
+        &mut self.queues
+    }
+
+    fn queue_events(&self) -> &[EventFd] {
+        &self.queue_evts
+    }
+
+    fn interrupt_evt(&self) -> &EventFd {
+        &self.interrupt_evt
+    }
+
+    fn interrupt_status(&self) -> Arc<AtomicUsize> {
+        self.interrupt_status.clone()
+    }
+
+    fn avail_features(&self) -> u64 {
+        self.avail_features
+    }
+
+    fn acked_features(&self) -> u64 {
+        self.acked_features
+    }
+
+    fn set_acked_features(&mut self, acked_features: u64) {
+        self.acked_features = acked_features;
+    }
+
+    fn read_config(&self, offset: u64, mut data: &mut [u8]) {
+        let config_space_bytes = self.config_space.as_slice();
+        let config_len = config_space_bytes.len() as u64;
+        if offset >= config_len {
+            error!("Failed to read config space");
+            return;
+        }
+
+        if let Some(end) = offset.checked_add(data.len() as u64) {
+            // This write can't fail, offset and end are checked against config_len.
+            data.write_all(
+                &config_space_bytes[offset as usize..cmp::min(end, config_len) as usize],
+            )
+            .unwrap();
+        }
+    }
+
+    fn write_config(&mut self, offset: u64, data: &[u8]) {
+        let data_len = data.len() as u64;
+        let config_space_bytes = self.config_space.as_mut_slice();
+        let config_len = config_space_bytes.len() as u64;
+        if offset + data_len > config_len {
+            error!("Failed to write config space");
+            return;
+        }
+        config_space_bytes[offset as usize..(offset + data_len) as usize].copy_from_slice(data);
+    }
+
+    fn is_activated(&self) -> bool {
+        match self.device_state {
+            DeviceState::Inactive => false,
+            DeviceState::Activated(_) => true,
+        }
+    }
+
+    fn activate(&mut self, mem: GuestMemoryMmap) -> ActivateResult {
+        if self.activate_evt.write(1).is_err() {
+            error!("Balloon: Cannot write to activate_evt");
+            METRICS.balloon.activate_fails.inc();
+            return Err(super::super::ActivateError::BadActivate);
+        }
+        self.device_state = DeviceState::Activated(mem);
```

Comment:
I think there is a possible race condition if the vCPU thread sleeps right before this line and the emulation thread attempts to do some work. This will lead to a panic via `DeviceState::Inactive => unreachable!(),`. I know this is highly unlikely as the activation part is driven by only one CPU in the guest, but as a best practice we should set the `device_state` before sending the activation notification.

It looks like this is so for all devices so it is worth digging deeper.

Reply:
The device is only accessed through an `Arc<Mutex>` so there is no race condition here. A lot of logic internal to the device relies on exclusive access of `self`.

Reply:
> The device is only accessed through an `Arc<Mutex>` so there is no race condition here. A lot of logic internal to the device relies on exclusive access of `self`.

Yes, you are right, but still I would prefer swapping the 2 statements.